File: packages/runar-compiler/src/optimizer/peephole.ts
/**
 * Peephole optimizer — runs on Stack IR before emission.
 *
 * Scans for short sequences of stack operations that can be replaced with
 * fewer or cheaper opcodes. Applies rules iteratively until a fixed point
 * is reached (no more changes).
 */

import type { StackOp, PushOp, OpcodeOp } from '../ir/index.js';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function isPush(op: StackOp): op is PushOp {
  return op.op === 'push';
}

function isOpcode(op: StackOp, code?: string): op is OpcodeOp {
  if (op.op !== 'opcode') return false;
  if (code !== undefined) return op.code === code;
  return true;
}

function isPushBigInt(op: StackOp, n: bigint): boolean {
  return isPush(op) && typeof op.value === 'bigint' && op.value === n;
}

function isPushZero(op: StackOp): boolean {
  return isPushBigInt(op, 0n);
}

function isPushOne(op: StackOp): boolean {
  return isPushBigInt(op, 1n);
}

// ---------------------------------------------------------------------------
// Peephole rules
// ---------------------------------------------------------------------------

/** A peephole rule: matches a window of ops and returns a replacement (or null). */
interface PeepholeRule {
  /** Number of ops this rule inspects. */
  windowSize: number;
  /** Try to match and return replacement ops, or null if no match. */
  match(ops: StackOp[]): StackOp[] | null;
}

const rules: PeepholeRule[] = [
  // -------------------------------------------------------------------------
  // Dead value elimination: PUSH x, DROP → remove both
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isPush(ops[0]!) && ops[1]!.op === 'drop') {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // DUP, DROP → remove both
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (ops[0]!.op === 'dup' && ops[1]!.op === 'drop') {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // SWAP, SWAP → remove both (identity)
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (ops[0]!.op === 'swap' && ops[1]!.op === 'swap') {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // PUSH 1, OP_ADD → OP_1ADD
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isPushOne(ops[0]!) && isOpcode(ops[1]!, 'OP_ADD')) {
        return [{ op: 'opcode', code: 'OP_1ADD' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // PUSH 1, OP_SUB → OP_1SUB
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isPushOne(ops[0]!) && isOpcode(ops[1]!, 'OP_SUB')) {
        return [{ op: 'opcode', code: 'OP_1SUB' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // PUSH 0, OP_ADD → remove both (identity: x + 0 = x)
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isPushZero(ops[0]!) && isOpcode(ops[1]!, 'OP_ADD')) {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // PUSH 0, OP_SUB → remove both (identity: x - 0 = x)
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isPushZero(ops[0]!) && isOpcode(ops[1]!, 'OP_SUB')) {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_NOT, OP_NOT → remove both (double negation)
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_NOT') && isOpcode(ops[1]!, 'OP_NOT')) {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_NEGATE, OP_NEGATE → remove both (double negation)
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_NEGATE') && isOpcode(ops[1]!, 'OP_NEGATE')) {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_EQUAL, OP_VERIFY → OP_EQUALVERIFY
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_EQUAL') && isOpcode(ops[1]!, 'OP_VERIFY')) {
        return [{ op: 'opcode', code: 'OP_EQUALVERIFY' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_CHECKSIG, OP_VERIFY → OP_CHECKSIGVERIFY
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_CHECKSIG') && isOpcode(ops[1]!, 'OP_VERIFY')) {
        return [{ op: 'opcode', code: 'OP_CHECKSIGVERIFY' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_NUMEQUAL, OP_VERIFY → OP_NUMEQUALVERIFY
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_NUMEQUAL') && isOpcode(ops[1]!, 'OP_VERIFY')) {
        return [{ op: 'opcode', code: 'OP_NUMEQUALVERIFY' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_CHECKMULTISIG, OP_VERIFY → OP_CHECKMULTISIGVERIFY
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_CHECKMULTISIG') && isOpcode(ops[1]!, 'OP_VERIFY')) {
        return [{ op: 'opcode', code: 'OP_CHECKMULTISIGVERIFY' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_DUP, OP_DROP → remove both (but not if DUP is needed elsewhere)
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_DUP') && isOpcode(ops[1]!, 'OP_DROP')) {
        return [];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // PUSH x, OP_DROP → remove both (same as generic push/drop but for opcode pushes)
  // -------------------------------------------------------------------------
  // Already covered by the first rule.

  // -------------------------------------------------------------------------
  // OP_OVER, OP_OVER → OP_2DUP
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (ops[0]!.op === 'over' && ops[1]!.op === 'over') {
        return [{ op: 'opcode', code: 'OP_2DUP' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // OP_DROP, OP_DROP → OP_2DROP
  // -------------------------------------------------------------------------
  {
    windowSize: 2,
    match(ops) {
      if (ops[0]!.op === 'drop' && ops[1]!.op === 'drop') {
        return [{ op: 'opcode', code: 'OP_2DROP' }];
      }
      return null;
    },
  },

  // -------------------------------------------------------------------------
  // SWAP, ROT → equivalent to a single ROT, SWAP in some patterns.
  // This is left as a future optimization.
  // -------------------------------------------------------------------------

  // =========================================================================
  // New 2-op rules: Roll/Pick depth simplification
  // The stack lowerer emits PUSH(depth) + Roll/Pick pairs. These rules
  // consume both the push and the roll/pick, replacing with a single opcode.
  // =========================================================================

  // PUSH 0, Roll{0} → remove both (no-op)
  {
    windowSize: 2,
    match(ops) {
      if (isPushBigInt(ops[0]!, 0n) && ops[1]!.op === 'roll' && ops[1]!.depth === 0) return [];
      return null;
    },
  },

  // PUSH 1, Roll{1} → Swap
  {
    windowSize: 2,
    match(ops) {
      if (isPushBigInt(ops[0]!, 1n) && ops[1]!.op === 'roll' && ops[1]!.depth === 1) return [{ op: 'swap' }];
      return null;
    },
  },

  // PUSH 2, Roll{2} → Rot
  {
    windowSize: 2,
    match(ops) {
      if (isPushBigInt(ops[0]!, 2n) && ops[1]!.op === 'roll' && ops[1]!.depth === 2) return [{ op: 'rot' }];
      return null;
    },
  },

  // PUSH 0, Pick{0} → Dup
  {
    windowSize: 2,
    match(ops) {
      if (isPushBigInt(ops[0]!, 0n) && ops[1]!.op === 'pick' && ops[1]!.depth === 0) return [{ op: 'dup' }];
      return null;
    },
  },

  // PUSH 1, Pick{1} → Over
  {
    windowSize: 2,
    match(ops) {
      if (isPushBigInt(ops[0]!, 1n) && ops[1]!.op === 'pick' && ops[1]!.depth === 1) return [{ op: 'over' }];
      return null;
    },
  },

  // =========================================================================
  // New 2-op rules
  // =========================================================================

  // SHA256, SHA256 → HASH256
  {
    windowSize: 2,
    match(ops) {
      if (isOpcode(ops[0]!, 'OP_SHA256') && isOpcode(ops[1]!, 'OP_SHA256')) {
        return [{ op: 'opcode', code: 'OP_HASH256' }];
      }
      return null;
    },
  },

  // PUSH 0, NUMEQUAL → NOT
  {
    windowSize: 2,
    match(ops) {
      if (isPushZero(ops[0]!) && isOpcode(ops[1]!, 'OP_NUMEQUAL')) {
        return [{ op: 'opcode', code: 'OP_NOT' }];
      }
      return null;
    },
  },

  // NOTE: PUSH 1, MUL identity rule is intentionally omitted.
  // In Bitcoin Script, PUSH 1 OP_MUL forces byte-to-number coercion.
  // SLH-DSA codegen relies on this behavior.

  // =========================================================================
  // New 3-op rules: constant folding
  // =========================================================================

  // PUSH(a), PUSH(b), ADD → PUSH(a+b)
  {
    windowSize: 3,
    match(ops) {
      if (isPush(ops[0]!) && typeof ops[0]!.value === 'bigint' &&
          isPush(ops[1]!) && typeof ops[1]!.value === 'bigint' &&
          isOpcode(ops[2]!, 'OP_ADD')) {
        const a = (ops[0]! as PushOp).value as bigint;
        const b = (ops[1]! as PushOp).value as bigint;
        return [{ op: 'push', value: a + b }];
      }
      return null;
    },
  },

  // PUSH(a), PUSH(b), SUB → PUSH(a-b)
  {
    windowSize: 3,
    match(ops) {
      if (isPush(ops[0]!) && typeof ops[0]!.value === 'bigint' &&
          isPush(ops[1]!) && typeof ops[1]!.value === 'bigint' &&
          isOpcode(ops[2]!, 'OP_SUB')) {
        const a = (ops[0]! as PushOp).value as bigint;
        const b = (ops[1]! as PushOp).value as bigint;
        return [{ op: 'push', value: a - b }];
      }
      return null;
    },
  },

  // PUSH(a), PUSH(b), MUL → PUSH(a*b)
  {
    windowSize: 3,
    match(ops) {
      if (isPush(ops[0]!) && typeof ops[0]!.value === 'bigint' &&
          isPush(ops[1]!) && typeof ops[1]!.value === 'bigint' &&
          isOpcode(ops[2]!, 'OP_MUL')) {
        const a = (ops[0]! as PushOp).value as bigint;
        const b = (ops[1]! as PushOp).value as bigint;
        return [{ op: 'push', value: a * b }];
      }
      return null;
    },
  },

  // =========================================================================
  // New 4-op rules: chain folding
  // =========================================================================

  // PUSH(a), ADD, PUSH(b), ADD → PUSH(a+b), ADD
  {
    windowSize: 4,
    match(ops) {
      if (isPush(ops[0]!) && typeof ops[0]!.value === 'bigint' &&
          isOpcode(ops[1]!, 'OP_ADD') &&
          isPush(ops[2]!) && typeof ops[2]!.value === 'bigint' &&
          isOpcode(ops[3]!, 'OP_ADD')) {
        const a = (ops[0]! as PushOp).value as bigint;
        const b = (ops[2]! as PushOp).value as bigint;
        return [{ op: 'push', value: a + b }, { op: 'opcode', code: 'OP_ADD' }];
      }
      return null;
    },
  },

  // PUSH(a), SUB, PUSH(b), SUB → PUSH(a+b), SUB
  {
    windowSize: 4,
    match(ops) {
      if (isPush(ops[0]!) && typeof ops[0]!.value === 'bigint' &&
          isOpcode(ops[1]!, 'OP_SUB') &&
          isPush(ops[2]!) && typeof ops[2]!.value === 'bigint' &&
          isOpcode(ops[3]!, 'OP_SUB')) {
        const a = (ops[0]! as PushOp).value as bigint;
        const b = (ops[2]! as PushOp).value as bigint;
        return [{ op: 'push', value: a + b }, { op: 'opcode', code: 'OP_SUB' }];
      }
      return null;
    },
  },
];

// ---------------------------------------------------------------------------
// Peephole optimizer entry point
// ---------------------------------------------------------------------------

/**
 * Apply peephole optimization rules to a list of stack ops.
 *
 * Rules are applied in a single left-to-right pass, then the entire pass
 * is repeated until no more changes occur (fixed-point iteration).
 *
 * If-ops are recursively optimized: the then/else branches are each
 * optimized independently.
 */
export function optimizeStackIR(ops: StackOp[]): StackOp[] {
  // First, recursively optimize nested if-blocks
  let current = ops.map(op => optimizeNestedIf(op));

  const MAX_ITERATIONS = 100;
  let iteration = 0;

  while (iteration < MAX_ITERATIONS) {
    const result = applyOnePass(current);
    if (!result.changed) break;
    current = result.ops;
    iteration++;
  }

  return current;
}

/**
 * Recursively optimize if-op branches.
 */
function optimizeNestedIf(op: StackOp): StackOp {
  if (op.op === 'if') {
    const optimizedThen = optimizeStackIR(op.then);
    const optimizedElse = op.else ? optimizeStackIR(op.else) : undefined;
    return {
      op: 'if',
      then: optimizedThen,
      else: optimizedElse,
    };
  }
  return op;
}

/**
 * Apply all peephole rules in a single left-to-right scan.
 * Tries wider windows first (greedy) for maximum reduction.
 */
function applyOnePass(ops: StackOp[]): { ops: StackOp[]; changed: boolean } {
  const result: StackOp[] = [];
  let changed = false;
  let i = 0;

  // Organize rules by window size for efficient matching
  const rulesBySize = new Map<number, PeepholeRule[]>();
  for (const rule of rules) {
    const list = rulesBySize.get(rule.windowSize) ?? [];
    list.push(rule);
    rulesBySize.set(rule.windowSize, list);
  }
  const windowSizes = [...rulesBySize.keys()].sort((a, b) => b - a); // largest first

  while (i < ops.length) {
    // Altstack round-trip elimination is disabled for now.
    // The safety check (net stack effect = 0) is necessary but not sufficient:
    // removing DUP+TOALTSTACK shifts items on the main stack, which invalidates
    // PICK/ROLL depths in the middle ops. Enabling this requires full stack
    // depth simulation which is future work.

    let matched = false;

    // Try rules from largest window to smallest
    for (const size of windowSizes) {
      if (i + size > ops.length) continue;

      const sizeRules = rulesBySize.get(size)!;
      const window = ops.slice(i, i + size);

      for (const rule of sizeRules) {
        const replacement = rule.match(window);
        if (replacement !== null) {
          result.push(...replacement);
          i += size;
          changed = true;
          matched = true;
          break;
        }
      }

      if (matched) break;
    }

    if (!matched) {
      result.push(ops[i]!);
      i++;
    }
  }

  return { ops: result, changed };
}

// ---------------------------------------------------------------------------
// Altstack round-trip elimination — DISABLED pending full stack simulation.
// See plan for details on implementing safe DUP TOALTSTACK <middle> FROMALTSTACK
// elimination. Requires precise stack depth tracking to avoid corrupting
// SLH-DSA and EC codegen patterns.
// ---------------------------------------------------------------------------

File: compilers/java/src/main/java/runar/compiler/passes/Peephole.java
package runar.compiler.passes;

import java.math.BigInteger;
import java.util.ArrayList;
import java.util.List;
import runar.compiler.ir.stack.BigIntPushValue;
import runar.compiler.ir.stack.DropOp;
import runar.compiler.ir.stack.DupOp;
import runar.compiler.ir.stack.IfOp;
import runar.compiler.ir.stack.OpcodeOp;
import runar.compiler.ir.stack.OverOp;
import runar.compiler.ir.stack.PickOp;
import runar.compiler.ir.stack.PushOp;
import runar.compiler.ir.stack.PushValue;
import runar.compiler.ir.stack.RollOp;
import runar.compiler.ir.stack.RotOp;
import runar.compiler.ir.stack.StackMethod;
import runar.compiler.ir.stack.StackOp;
import runar.compiler.ir.stack.StackProgram;
import runar.compiler.ir.stack.SwapOp;

/**
 * Peephole optimizer — runs on Stack IR before emission.
 *
 * <p>Port of {@code packages/runar-compiler/src/optimizer/peephole.ts}.
 * Rules are applied iteratively to a fixed point, with wider windows tried
 * first so adjacent rewrites compose correctly. If-branches are optimised
 * recursively.
 */
public final class Peephole {

    private Peephole() {}

    private static final int MAX_ITERATIONS = 100;

    public static StackProgram run(StackProgram program) {
        List<StackMethod> optimised = new ArrayList<>(program.methods().size());
        for (StackMethod m : program.methods()) {
            List<StackOp> ops = optimize(m.ops());
            optimised.add(new StackMethod(m.name(), ops, m.maxStackDepth()));
        }
        return new StackProgram(program.contractName(), optimised);
    }

    public static List<StackOp> optimize(List<StackOp> ops) {
        List<StackOp> current = new ArrayList<>(ops.size());
        for (StackOp op : ops) current.add(optimizeNestedIf(op));

        for (int it = 0; it < MAX_ITERATIONS; it++) {
            Pass pass = applyOnePass(current);
            if (!pass.changed) break;
            current = pass.ops;
        }
        return current;
    }

    private static StackOp optimizeNestedIf(StackOp op) {
        if (op instanceof IfOp i) {
            List<StackOp> thenB = optimize(i.thenBranch());
            List<StackOp> elseB = i.elseBranch() != null ? optimize(i.elseBranch()) : null;
            if (elseB == null) return new IfOp(thenB);
            return new IfOp(thenB, elseB);
        }
        return op;
    }

    private static final class Pass {
        final List<StackOp> ops;
        final boolean changed;
        Pass(List<StackOp> ops, boolean changed) {
            this.ops = ops;
            this.changed = changed;
        }
    }

    private static Pass applyOnePass(List<StackOp> ops) {
        List<StackOp> out = new ArrayList<>();
        boolean changed = false;
        int i = 0;
        while (i < ops.size()) {
            // Widest window first: 4, 3, then 2.
            if (i + 4 <= ops.size()) {
                List<StackOp> r = matchWindow4(ops.get(i), ops.get(i + 1), ops.get(i + 2), ops.get(i + 3));
                if (r != null) {
                    out.addAll(r);
                    i += 4;
                    changed = true;
                    continue;
                }
            }
            if (i + 3 <= ops.size()) {
                List<StackOp> r = matchWindow3(ops.get(i), ops.get(i + 1), ops.get(i + 2));
                if (r != null) {
                    out.addAll(r);
                    i += 3;
                    changed = true;
                    continue;
                }
            }
            if (i + 2 <= ops.size()) {
                List<StackOp> r = matchWindow2(ops.get(i), ops.get(i + 1));
                if (r != null) {
                    out.addAll(r);
                    i += 2;
                    changed = true;
                    continue;
                }
            }
            out.add(ops.get(i));
            i++;
        }
        return new Pass(out, changed);
    }

    // ---------- window-2 rules ----------

    static List<StackOp> matchWindow2(StackOp a, StackOp b) {
        // PUSH x, DROP → remove both (dead value elimination)
        if (a instanceof PushOp && b instanceof DropOp) return List.of();

        // DUP, DROP → remove both
        if (a instanceof DupOp && b instanceof DropOp) return List.of();

        // SWAP, SWAP → identity
        if (a instanceof SwapOp && b instanceof SwapOp) return List.of();

        // PUSH 1, OP_ADD → OP_1ADD
        if (isPushBigInt(a, 1) && isOpcode(b, "OP_ADD")) return List.of(new OpcodeOp("OP_1ADD"));
        // PUSH 1, OP_SUB → OP_1SUB
        if (isPushBigInt(a, 1) && isOpcode(b, "OP_SUB")) return List.of(new OpcodeOp("OP_1SUB"));
        // PUSH 0, OP_ADD → identity
        if (isPushBigInt(a, 0) && isOpcode(b, "OP_ADD")) return List.of();
        // PUSH 0, OP_SUB → identity
        if (isPushBigInt(a, 0) && isOpcode(b, "OP_SUB")) return List.of();

        // OP_NOT, OP_NOT → remove both
        if (isOpcode(a, "OP_NOT") && isOpcode(b, "OP_NOT")) return List.of();
        // OP_NEGATE, OP_NEGATE → remove both
        if (isOpcode(a, "OP_NEGATE") && isOpcode(b, "OP_NEGATE")) return List.of();

        // OP_EQUAL + OP_VERIFY → OP_EQUALVERIFY
        if (isOpcode(a, "OP_EQUAL") && isOpcode(b, "OP_VERIFY")) return List.of(new OpcodeOp("OP_EQUALVERIFY"));
        // OP_CHECKSIG + OP_VERIFY → OP_CHECKSIGVERIFY
        if (isOpcode(a, "OP_CHECKSIG") && isOpcode(b, "OP_VERIFY")) return List.of(new OpcodeOp("OP_CHECKSIGVERIFY"));
        // OP_NUMEQUAL + OP_VERIFY → OP_NUMEQUALVERIFY
        if (isOpcode(a, "OP_NUMEQUAL") && isOpcode(b, "OP_VERIFY")) return List.of(new OpcodeOp("OP_NUMEQUALVERIFY"));
        // OP_CHECKMULTISIG + OP_VERIFY → OP_CHECKMULTISIGVERIFY
        if (isOpcode(a, "OP_CHECKMULTISIG") && isOpcode(b, "OP_VERIFY")) return List.of(new OpcodeOp("OP_CHECKMULTISIGVERIFY"));

        // OP_DUP + OP_DROP → remove both
        if (isOpcode(a, "OP_DUP") && isOpcode(b, "OP_DROP")) return List.of();

        // OVER + OVER → OP_2DUP
        if (a instanceof OverOp && b instanceof OverOp) return List.of(new OpcodeOp("OP_2DUP"));
        // DROP + DROP → OP_2DROP
        if (a instanceof DropOp && b instanceof DropOp) return List.of(new OpcodeOp("OP_2DROP"));

        // PUSH 0 + Roll{0} → remove both
        if (isPushBigInt(a, 0) && isRoll(b, 0)) return List.of();
        // PUSH 1 + Roll{1} → SWAP
        if (isPushBigInt(a, 1) && isRoll(b, 1)) return List.of(new SwapOp());
        // PUSH 2 + Roll{2} → ROT
        if (isPushBigInt(a, 2) && isRoll(b, 2)) return List.of(new RotOp());
        // PUSH 0 + Pick{0} → DUP
        if (isPushBigInt(a, 0) && isPick(b, 0)) return List.of(new DupOp());
        // PUSH 1 + Pick{1} → OVER
        if (isPushBigInt(a, 1) && isPick(b, 1)) return List.of(new OverOp());

        // OP_SHA256 + OP_SHA256 → OP_HASH256
        if (isOpcode(a, "OP_SHA256") && isOpcode(b, "OP_SHA256")) return List.of(new OpcodeOp("OP_HASH256"));

        // PUSH 0 + OP_NUMEQUAL → OP_NOT
        if (isPushBigInt(a, 0) && isOpcode(b, "OP_NUMEQUAL")) return List.of(new OpcodeOp("OP_NOT"));

        return null;
    }

    // ---------- window-3 rules ----------

    static List<StackOp> matchWindow3(StackOp a, StackOp b, StackOp c) {
        BigInteger av = pushBigInt(a);
        BigInteger bv = pushBigInt(b);
        if (av != null && bv != null) {
            if (isOpcode(c, "OP_ADD")) return List.of(new PushOp(PushValue.of(av.add(bv))));
            if (isOpcode(c, "OP_SUB")) return List.of(new PushOp(PushValue.of(av.subtract(bv))));
            if (isOpcode(c, "OP_MUL")) return List.of(new PushOp(PushValue.of(av.multiply(bv))));
        }
        return null;
    }

    // ---------- window-4 rules ----------

    static List<StackOp> matchWindow4(StackOp a, StackOp b, StackOp c, StackOp d) {
        BigInteger av = pushBigInt(a);
        BigInteger cv = pushBigInt(c);
        if (av != null && cv != null) {
            if (isOpcode(b, "OP_ADD") && isOpcode(d, "OP_ADD")) {
                return List.of(new PushOp(PushValue.of(av.add(cv))), new OpcodeOp("OP_ADD"));
            }
            if (isOpcode(b, "OP_SUB") && isOpcode(d, "OP_SUB")) {
                return List.of(new PushOp(PushValue.of(av.add(cv))), new OpcodeOp("OP_SUB"));
            }
        }
        return null;
    }

    // ---------- predicates ----------

    private static boolean isOpcode(StackOp op, String code) {
        return op instanceof OpcodeOp o && code.equals(o.code());
    }

    private static boolean isPushBigInt(StackOp op, long n) {
        if (!(op instanceof PushOp p)) return false;
        if (!(p.value() instanceof BigIntPushValue bi)) return false;
        return bi.value().equals(BigInteger.valueOf(n));
    }

    private static BigInteger pushBigInt(StackOp op) {
        if (op instanceof PushOp p && p.value() instanceof BigIntPushValue bi) {
            return bi.value();
        }
        return null;
    }

    private static boolean isRoll(StackOp op, long depth) {
        return op instanceof RollOp r && r.depth().equals(BigInteger.valueOf(depth));
    }

    private static boolean isPick(StackOp op, long depth) {
        return op instanceof PickOp p && p.depth().equals(BigInteger.valueOf(depth));
    }
}
